File: packages/shared/services/memory_storage.py
import hashlib
import logging
import time
from typing import TypedDict

from qdrant_client import QdrantClient
from qdrant_client.models import (
    Distance,
    FieldCondition,
    Filter,
    MatchValue,
    Direction,
    OrderBy,
    PointIdsList,
    PointStruct,
    ScoredPoint,
    VectorParams,
)

from shared.config import config

class MemoryRecord(TypedDict):
    id: int
    text: str
    created_at: int | None


COLLECTION_NAME = "user_memory"
VECTOR_SIZE = config.embedding.vector_size


class MemoryStorage:
    def __init__(self) -> None:
        self.logger = logging.getLogger(self.__class__.__name__)
        self._client = QdrantClient(host=config.qdrant.host, port=config.qdrant.port)
        self._ensure_collection()

    def _ensure_collection(self) -> None:
        if self._client.collection_exists(collection_name=COLLECTION_NAME):
            return
        self._client.create_collection(
            collection_name=COLLECTION_NAME,
            vectors_config=VectorParams(size=VECTOR_SIZE, distance=Distance.COSINE),
        )
        self.logger.info("Created collection '%s'", COLLECTION_NAME)

    @staticmethod
    def _make_id(user_id: str, text: str) -> int:
        digest = hashlib.sha256(f"{user_id}:{text}".encode()).digest()
        return int.from_bytes(digest[:8], "big")

    @staticmethod
    def _user_filter(user_id: str) -> Filter:
        return Filter(must=[FieldCondition(key="user_id", match=MatchValue(value=user_id))])

    def save(self, user_id: str, text: str, vector: list[float]) -> int:
        point_id = self._make_id(user_id, text)
        self._client.upsert(
            collection_name=COLLECTION_NAME,
            points=[
                PointStruct(
                    id=point_id,
                    vector=vector,
                    payload={"user_id": user_id, "text": text, "created_at": int(time.time())},
                )
            ],
        )
        return point_id

    def search(self, user_id: str, vector: list[float], k: int = 5) -> list[ScoredPoint]:
        result = self._client.query_points(
            collection_name=COLLECTION_NAME,
            query=vector,
            query_filter=self._user_filter(user_id),
            limit=k,
        )
        return result.points

    def list_all(self, user_id: str) -> list[MemoryRecord]:
        result = self._client.scroll(
            collection_name=COLLECTION_NAME,
            scroll_filter=self._user_filter(user_id),
            order_by=OrderBy(key="created_at", direction=Direction.DESC),
            limit=100,
        )
        points = result[0]
        return [
            MemoryRecord(
                id=int(p.id),
                text=str((p.payload or {}).get("text", "")),
                created_at=(p.payload or {}).get("created_at"),
            )
            for p in points
        ]

    def delete(self, point_id: int) -> None:
        self._client.delete(
            collection_name=COLLECTION_NAME,
            points_selector=PointIdsList(points=[point_id]),
        )


memory_storage = MemoryStorage()

File: packages/chatbot/server/services/memory-storage.ts
import { randomUUID } from "node:crypto";

import { QdrantClient } from "@qdrant/qdrant-js";

import { config, qdrantCollections } from "@/server/config";
import { embedderService } from "@/server/services/embedder";

export type MemoryRecord = {
  id: string;
  text: string;
  createdAt: number;
};

export class MemoryStorage {
  private static instance: MemoryStorage | null = null;
  private readonly qdrant: QdrantClient;
  private readonly collectionName = qdrantCollections.memory;

  private constructor() {
    this.qdrant = new QdrantClient({ url: config.qdrant.url });
  }

  static getInstance(): MemoryStorage {
    if (!MemoryStorage.instance) {
      MemoryStorage.instance = new MemoryStorage();
    }
    return MemoryStorage.instance;
  }

  private async ensureCollection(): Promise<void> {
    const { exists } = await this.qdrant.collectionExists(this.collectionName);
    if (!exists) {
      await this.qdrant.createCollection(this.collectionName, {
        vectors: {
          size: config.embedding.vectorSize,
          distance: "Cosine",
        },
      });
    }
  }

  private userFilter(userId: string) {
    return {
      must: [
        {
          key: "userId",
          match: { value: userId },
        },
      ],
    };
  }

  async save(userId: string, text: string) {
    await this.ensureCollection();
    const vector = await embedderService.embedQuery(text);
    const id = randomUUID();

    await this.qdrant.upsert(this.collectionName, {
      wait: true,
      points: [
        {
          id,
          vector,
          payload: {
            userId,
            createdAt: Math.floor(Date.now() / 1000),
            text,
          },
        },
      ],
    });

    return id;
  }

  async search(
    userId: string,
    query: string,
    k = 5,
  ): Promise<Array<{ id: string; text: string; score: number | null }>> {
    await this.ensureCollection();
    const vector = await embedderService.embedQuery(query);

    const points = await this.qdrant.search(this.collectionName, {
      vector,
      filter: this.userFilter(userId),
      limit: k,
      with_payload: true,
    });

    return points.map((point) => ({
      id: String(point.id),
      text: String(point.payload?.text ?? ""),
      score: point.score ?? null,
    }));
  }

  async list(userId: string): Promise<MemoryRecord[]> {
    await this.ensureCollection();

    const data = await this.qdrant.scroll(this.collectionName, {
      filter: this.userFilter(userId),
      order_by: {
        key: "createdAt",
        direction: "desc",
      },
      with_payload: true,
      limit: 100,
    });

    return data.points.map((point) => ({
      id: String(point.id),
      text: String(point.payload?.text ?? ""),
      createdAt: Number(point.payload?.createdAt ?? 0),
    }));
  }

  async delete(memoryId: string): Promise<void> {
    await this.ensureCollection();
    await this.qdrant.delete(this.collectionName, {
      wait: true,
      points: [memoryId],
    });
  }
}

export const memoryStorage = MemoryStorage.getInstance();
